### PR TITLE
Fix for Criteo OOM

### PR DESCRIPTION
**Pull request by pomonam**
I rewrote the code for `dot interaction` and `embedding`. These changes fix the OOM issue during training, but not the OOM issue during evaluation (which is quite strange) I further reduced the batch size during evaluation to fix this. This PR makes the Criteo workload on PyTorch runnable, but further investigation would be needed to make the eval faster.

### CODE REVIEW

**Comment by priyakasimbeg** [comment]
Hi Juhan! This looks great, thank you sm for these fixes.
It seems like one of the tests is failing on the param shapes between jax and pytorch. It seems like this is intended though. @znado I'm not sure if we want to disable the test for Criteo?

**Comment by runame** [comment]
I haven't looked at this yet, but if @pomonam has changed the embedding implementation the param shapes test has to be adjusted so that the shape is matched correctly to the Jax implementation, similar to the changes I had to make in #489 to adjust for the new `MultiheadAttention` implementation.

**Comment by pomonam** [comment]
Thank you! I fixed the tests.

**Review by priyakasimbeg — APPROVED** [review]
LGTM

**Comment by priyakasimbeg** [comment]
I LGTMd this but will wait for @runame merge this in after you've had a chance to review it. 
Just one question, we initialize the embedding layer from uniform distribution, do you know by any chance if that's consistent with how it was initialized before this PR @pomonam? I noticed the _reset_parameters uses xavier_uniform or normal distributions for initializations.

**Comment by pomonam** [comment]
Yes, I think the initialization is identical to what was done before.

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/models.py
Comment (runame) [inline algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/models.py:67]: Nit: single quotation marks.

**Review by runame — COMMENTED** [review]
LGTM!

Just to make sure we should check for equivalence using the traindiffs test, but it might currently be broken due to the new embedding implementation. @chandramouli-sastry might be able to help with this.

**Comment by pomonam** [comment]
Yes, diff test would be great!

**Comment by pomonam** [comment]
Yes, diff test would be great!

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/models.py
Comment (pomonam) [reply algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/models.py]: Ah, thank you for the catch :). Fixed!

**Review by priyakasimbeg — APPROVED** [review]

**Pull request merged**
